src/GUI/GameMoveListener.java

```diff
@@ -0,0 +1,115 @@
+package GUI;
+
+import Cards.Card;
+import Gameplay.*;
+
+import javax.swing.event.MouseInputAdapter;
+import java.awt.*;
+import java.awt.event.MouseEvent;
+
+public class GameMoveListener extends MouseInputAdapter {
```

Comment:
Design wise comment:

A lot of logic based on:
if clicked on x, remember x, otherwise, if clicked on y, remember y, ....
This repeats, causes logic errors.

Advice: use the [Adapter](https://refactoring.guru/design-patterns/adapter) pattern.

How to: 
Keep a queue of pressed components on each mouse pressed.
When mouse is released, if two elements in queue, pop both and check using the adapter(s) if the move is legal.

You would have the following adapters:
`TableauToFoundationMoveAdapter` (can reverse)
`PileToFoundationMoveAdapter` (can reverse)
...
Or, even, abstract further
consider `from` and `to` as abstract destinations (foundation, pile, tableau, ...), and you have `AbstractMoveAdapter` taking instances of `from` and `to` to check move legality and apply side effects if legal.

Reply:
Considering the adapter suggestion, I will refrain from commenting on the logic of the existing class as it is, honestly, compliated, convoluted and headache inducing.